### PR TITLE
Add Node maintenance test cases

### PR DESCRIPTION
**Pull request by rhrazdil**
Added test cases for node maintenance.
ATM test cases are implemented with using pods instead of VMs due to lack of testing environment.https://polarion.engineering.redhat.com/polarion/#/project/CNV/wiki/Testing/_UI_%20Node%20Maintenante

### CODE REVIEW

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  const leakedResources = new Set<string>();
+  const podName = `maintenance-${testName}`;
+  const pod = new Pod(podName, testName);
+  let computeNodeHostname = '';
+  let computeNodeURL = '';
+
+  beforeAll(async() => {
+    // Create an example hello-world pod
+    execSync(`echo '${JSON.stringify(examplePod(podName, {}))}' | kubectl create -f -`);
+    leakedResources.add(JSON.stringify({name: podName, namespace: `${testName}`, kind: 'pods'}));
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: this object can be reused if examplePod params are changed

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  it('Terminates running pod on maintenance node', async() => {
+    await pod.navigateToTab(overviewTab);
+    await pod.waitForStatusIcon(podView.statusIcons.terminating, PAGE_LOAD_TIMEOUT);
+
+    await browser.get(`${appHost}/k8s/ns/${testName}/pods`);
+    await isLoaded();
+
+    await filterForName(podName);
+    await browser.wait(until.and(waitForCount(resourceRows, 0)), VM_BOOTUP_TIMEOUT);
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: `VM_BOOTUP_TIMEOUT`, shouldn't some other timeout be here ?

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts (hunk 1 above)
Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: \+ also used in `it()` functions

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import { examplePod } from './mocks';
+import { appHost, testName } from '../../protractor.conf';
+import { isLoaded, filterForName, resourceRows } from '../../views/crud.view';
+import { overviewTab } from '../../views/kubevirt/virtualMachine.view';
+import { detailViewAction } from '../../views/kubevirt/vm.actions.view';
+import Pod from './models/pod';
+import * as podView from '../../views/kubevirt/pod.view';
+
+
+describe('Test Node Maintenance Mode', () => {
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: shouldn't we also test the node status in `/k8s/cluster/nodes`?

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    leakedResources.add(JSON.stringify({name: `${pod.name}`, namespace: `${pod.namespace}`, kind: `${pod.kind}`}));
+    await pod.navigateToTab(overviewTab);
+
+    let errorMessage = '';
+    try {
+      await pod.waitForStatusIcon(podView.statusIcons.running, POD_START_TIMEOUT);
+    } catch (error) {
+      errorMessage = error.message;
+    }
+    expect(errorMessage).toBe('Time out error');
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: `Time out error` should be a constant

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    await pod.action('Delete');
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: I would rather check CLI here or in pod.action than wait for redirect. It is a bit safer.

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    leakedResources.delete(JSON.stringify({name: `${pod.name}`, namespace: `${pod.namespace}`, kind: `${pod.kind}`}));
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: mm, it would be probably more useful to create some function like this:

```ts
removeLeakedResource(leakedResources: Set<string>, resource: Object)
```

Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: in that case removeLeakedResources should be better named. `assertNoLeakedResources` ?

Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: I think we could use functions for adding/removing a resource in to the a Set, however, assertNoLeakedResources is a bit misleading. The point is not asserting that nothing leaked, but having a safeguard if something breaks during the execution (in which case a potential delete action at the end of a spec wouldn't be executed)

Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: done.

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts (hunk 5 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: Changed this so that it's being removed by cli as it's not really necessary to remove the pod in UI + improved the delete action so that it waits until pod actually disappears in UI

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts (hunk 4 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: done

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts (hunk 3 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: Added

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts (hunk 2 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: done

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts (hunk 1 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: done

File: frontend/integration-tests/tests/kubevirt/mocks.ts [hunk 7]
```diff
@@ -289,6 +289,33 @@ export const cloudInitCustomScriptConfig = {
   customScript: basicVmConfig.cloudInitScript,
 };
 
+export const examplePod = (name: string, nodeSelector: any) => {
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/mocks.ts]: Agee, I have also started to realize that the parameters are getting cumbersome

File: frontend/integration-tests/tests/kubevirt/utils/utils.ts [hunk 8]
```diff
@@ -47,12 +40,33 @@ export function removeLeakedResources(leakedResources: Set<string>) {
...
+    JSON.stringify({name: resource.metadata.name, namespace: resource.metadata.namespace, kind: resource.kind})
+  );
+}
+
+export function createResource(resource) {
+  execSync(`echo '${JSON.stringify(resource)}' | kubectl create -f -`);
+}
+
 export function createResources(resources) {
   resources.forEach(resource => {
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/utils/utils.ts]: we can call createResource forEach resource

File: frontend/integration-tests/tests/kubevirt/utils/utils.ts [hunk 9]
```diff
@@ -47,12 +40,33 @@ export function removeLeakedResources(leakedResources: Set<string>) {
...
     execSync(`echo '${JSON.stringify(resource)}' | kubectl create -f -`);
   });
 }
 
+export function deleteResource(resource) {
+  const kind = resource.kind === 'NetworkAttachmentDefinition' ? 'net-attach-def' : resource.kind;
+  execSync(`kubectl delete -n ${resource.metadata.namespace} --cascade ${kind} ${resource.metadata.name}`);
+}
+
 export function deleteResources(resources) {
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/utils/utils.ts]: same here

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+    await filterForName(podName);
+    await browser.wait(until.and(waitForCount(resourceRows, 0)), POD_TERMINATION_TIMEOUT);
+    removeLeakableResource(leakedResources, podResource);(podResource);
+  }, POD_CREATE_DELETE_TIMEOUT);
+
+  it('Node in Maintenance mode is not schedulable', async() => {
+    podResourceOpts.nodeSelector = {'kubernetes.io/hostname': computeNodeHostname};
+    createResource(examplePod(podResourceOpts));
+    addLeakableResource(leakedResources, podResource);(podResource);
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: leftover?

File: frontend/integration-tests/tests/kubevirt/mocks.ts [hunk 11]
```diff
@@ -289,6 +289,33 @@ export const cloudInitCustomScriptConfig = {
   customScript: basicVmConfig.cloudInitScript,
 };
 
+export const examplePod = (options) => {
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/mocks.ts]: it is a bit easier to read if you use object deconstruction here

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts (hunk 6 above)
Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: ah, ok

File: frontend/integration-tests/tests/kubevirt/utils/utils.ts (hunk 8 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/utils/utils.ts]: Right

File: frontend/integration-tests/tests/kubevirt/utils/utils.ts (hunk 9 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/utils/utils.ts]: done

File: frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts (hunk 10 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/node.maintenance.scenario.ts]: ups

File: frontend/integration-tests/tests/kubevirt/mocks.ts (hunk 11 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/mocks.ts]: done

File: frontend/integration-tests/tests/kubevirt/utils/utils.ts (hunk 8 above)
Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/utils/utils.ts]: no need to use lambda

File: frontend/integration-tests/tests/kubevirt/utils/utils.ts [hunk 12]
```diff
@@ -47,16 +40,37 @@ export function removeLeakedResources(leakedResources: Set<string>) {
...
   resources.forEach(resource => {
-    const kind = resource.kind === 'NetworkAttachmentDefinition' ? 'net-attach-def' : resource.kind;
-    execSync(`kubectl delete -n ${resource.metadata.namespace} --cascade ${kind} ${resource.metadata.name}`);
+    resource.kind = resource.kind === 'NetworkAttachmentDefinition' ? 'net-attach-def' : resource.kind;
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/utils/utils.ts]: why is this line needed here? Isn't this also executed in `deleteResource` ?

Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/utils/utils.ts]: Is this what you mean in the previous comment? `resources.forEach(resource => createResource(resource));`

And yes here it is not necessary.

Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/utils/utils.ts]: you mean the lamdas? 
```js
resources.forEach(createResource)
```

Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/utils/utils.ts]: Ah, cool, didn't realize it works like that. Done.
